src/backend/distributed/commands/index.c

```diff
@@ -120,103 +127,121 @@ List *
 PreprocessIndexStmt(Node *node, const char *createIndexCommand)
 {
 	IndexStmt *createIndexStatement = castNode(IndexStmt, node);
-	List *ddlJobs = NIL;
+
+	RangeVar *relationRangeVar = createIndexStatement->relation;
+	if (relationRangeVar == NULL)
+	{
+		/* let's be on the safe side */
+		return NIL;
+	}
 
 	/*
-	 * We first check whether a distributed relation is affected. For that, we need to
-	 * open the relation. To prevent race conditions with later lookups, lock the table,
-	 * and modify the rangevar to include the schema.
+	 * We first check whether a distributed relation is affected. For that,
+	 * we need to open the relation. To prevent race conditions with later
+	 * lookups, lock the table.
+	 *
+	 * XXX: Consider using RangeVarGetRelidExtended with a permission
+	 * checking callback. Right now we'll acquire the lock before having
+	 * checked permissions, and will only fail when executing the actual
+	 * index statements.
 	 */
-	if (createIndexStatement->relation != NULL)
+	LOCKMODE lockMode = GetCreateIndexRelationLockMode(createIndexStatement);
+	Relation relation = table_openrv(relationRangeVar, lockMode);
```

Comment:
I wonder whether IF NOT EXISTS takes an aggressive lock too, but not really this PR's concern.